compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirAnnotationArgumentChecker.kt

```diff
@@ -14,19 +13,17 @@ import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
 import org.jetbrains.kotlin.fir.analysis.diagnostics.FirDiagnosticFactory0
 import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
 import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
-import org.jetbrains.kotlin.fir.declarations.FirDeclaration
 import org.jetbrains.kotlin.fir.expressions.*
 import org.jetbrains.kotlin.psi.KtExpression
 
-object FirAnnotationArgumentChecker : FirBasicDeclarationChecker() {
-    override fun check(declaration: FirDeclaration, context: CheckerContext, reporter: DiagnosticReporter) {
-        if (declaration !is FirAnnotationContainer) return
-        for (declarationOfAnnotation in declaration.annotations) {
-            for ((arg, _) in declarationOfAnnotation.argumentMapping ?: continue) {
-                val expression = (arg as? FirNamedArgumentExpression)?.expression ?: arg
+object FirAnnotationArgumentChecker : FirAnnotationCallChecker() {
+    override fun check(expression: FirAnnotationCall, context: CheckerContext, reporter: DiagnosticReporter) {
+        expression.argumentMapping?.let {
+            for ((arg, _) in it) {
+                val argExpression = (arg as? FirNamedArgumentExpression)?.expression ?: arg
 
-                checkAnnotationArgumentWithSubElements(expression, context.session, reporter, context)
-                    ?.let { reporter.reportOn(expression.source, it, context) }
+                checkAnnotationArgumentWithSubElements(argExpression, context.session, reporter, context)
+                    ?.let { reporter.reportOn(argExpression.source, it, context) }
```

Comment:
It is out of scope of this PR, and not sure you will keep working on this diagnostic, but according to [this](https://punzki.github.io/Kotlin-DiagnosticTestsDiff/2021-05-05/diffs/tests/varargs/assignNonConstSingleArrayElementAsVarargInAnnotationError.kt.html), I guess this source needs some fixes.

Reply:
Thanks for pointing that out (I should use my own data :slightly_smiling_face:). I think those are fixes on source positioning side.